sdk/src/azure/platform/az_posix.c

```diff
@@ -26,3 +40,124 @@ AZ_NODISCARD az_result az_platform_sleep_msec(int32_t milliseconds)
   (void)usleep((useconds_t)milliseconds * _az_TIME_MICROSECONDS_PER_MILLISECOND);
   return AZ_OK;
 }
+
+AZ_NODISCARD az_result az_platform_get_random(int32_t* out_random)
+{
+  _az_PRECONDITION_NOT_NULL(out_random);
+  
+  *out_random = (int32_t)random();
+  return AZ_OK;
+}
+
+void _timer_callback_handler(union sigval sv)
+{
+  az_platform_timer *timer_handle = sv.sival_ptr;
+
+  _az_PRECONDITION_NOT_NULL(timer_handle);
+  _az_PRECONDITION_NOT_NULL(timer_handle->platform_timer._internal.callback);
+
+  timer_handle->platform_timer._internal.callback(timer_handle->platform_timer._internal.sdk_data);
+}
+
+AZ_NODISCARD az_result az_platform_timer_create(
+    az_platform_timer* timer_handle,
+    az_platform_timer_callback callback,
+    void* sdk_data)
+{
+  _az_PRECONDITION_NOT_NULL(timer_handle);
+  _az_PRECONDITION_NOT_NULL(callback);
+  memset(timer_handle, 0, sizeof(az_platform_timer));
+
+  timer_handle->platform_timer._internal.callback = callback;
+  timer_handle->platform_timer._internal.sdk_data = sdk_data;
+
+  timer_handle->_internal.sev.sigev_notify = SIGEV_THREAD;
+  timer_handle->_internal.sev.sigev_notify_function = &_timer_callback_handler;
+  timer_handle->_internal.sev.sigev_value.sival_ptr = timer_handle;
+
+  if (0 != timer_create(
+          CLOCK_MONOTONIC, &timer_handle->_internal.sev, &timer_handle->_internal.timerid))
+  {
+    return AZ_ERROR_ARG;
+  }
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result
+az_platform_timer_start(az_platform_timer* timer_handle, int32_t milliseconds)
+{
+  _az_PRECONDITION_NOT_NULL(timer_handle);
+
+  timer_handle->_internal.trigger.it_value.tv_sec = milliseconds / _az_TIME_MILLISECONDS_PER_SECOND;
+  timer_handle->_internal.trigger.it_value.tv_nsec
+      = (milliseconds % _az_TIME_MILLISECONDS_PER_SECOND) * _az_TIME_NANOSECONDS_PER_MILLISECOND;
+
+  if (0 != timer_settime(
+          timer_handle->_internal.timerid, 0, &timer_handle->_internal.trigger, NULL))
+  {
+    return AZ_ERROR_ARG;
+  }
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_platform_timer_destroy(az_platform_timer* timer_handle)
+{
+  _az_PRECONDITION_NOT_NULL(timer_handle);
+
+  if (0 != timer_delete(timer_handle->_internal.timerid))
+  {
+    return AZ_ERROR_ARG;
+  }
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_platform_mutex_init(az_platform_mutex* mutex_handle)
+{
+  _az_PRECONDITION_NOT_NULL(mutex_handle);
+
+  if (0 != pthread_mutex_init(mutex_handle, NULL))
```

Comment:
This is incorrect within the new design. The mutex must be re-entrant.

I've updated the code in my latest iteration of the proof-of-concept: https://github.com/CIPop/azure-sdk-for-c/blob/hfsm/sdk/src/azure/platform/az_posix.c#L96